Package.swift

```diff
@@ -1,20 +1,45 @@
+// swift-tools-version:4.0
 import PackageDescription
 
 let package = Package(
   name: "HaCWebsite",
-  targets: [
-    Target(name: "HaCWebsiteLib", dependencies: ["HaCTML"]),
-    Target(name: "HaCWebsite", dependencies: ["HaCWebsiteLib"]),
-    Target(name: "HaCTML")
+  products: [
+    .library(name: "HaCTML", targets: ["HaCTML"]),
+    .library(name: "HaCWebsiteLib", targets: ["HaCWebsiteLib"]),
+    .executable(name: "HaCWebsite", targets: ["HaCWebsite"])
   ],
   dependencies: [
-    .Package(url: "https://github.com/IBM-Swift/Kitura.git", majorVersion: 1, minor: 6),
-    .Package(url: "https://github.com/IBM-Swift/HeliumLogger.git", majorVersion: 1, minor: 6),
-    .Package(url: "https://github.com/IBM-Swift/Kitura-Markdown.git", majorVersion: 0),
-    .Package(url: "https://github.com/hackersatcambridge/YamlSwift.git", majorVersion: 4, minor: 0),
-    .Package(url: "https://github.com/jaredkhan/SwiftDotEnv.git", majorVersion: 1, minor: 2),
-    .Package(url: "https://github.com/alexaubry/HTMLString", majorVersion: 3, minor: 0),
-    .Package(url: "https://github.com/vapor/fluent.git", majorVersion: 2, minor: 3),
-    .Package(url: "https://github.com/vapor-community/postgresql-driver.git", majorVersion: 2, minor: 0)
-  ]
-)
+    .package(url: "https://github.com/IBM-Swift/Kitura.git", .upToNextMinor(from: "1.7.9")),
+    .package(url: "https://github.com/IBM-Swift/HeliumLogger.git", .upToNextMinor(from: "1.7.1")),
```

Comment:
What does `.upToNextMinor` mean? Is it basically saying, we are happy with `1.7.*`>`1.7.9`. If one of these changes, will that change `Package.resolved`?

Reply:
Ok, interestingly, when I ran locally, my Package.resolved did change, a lot:

```git
diff --git a/Package.resolved b/Package.resolved
index 2f8988d..96b13fb 100644
--- a/Package.resolved
+++ b/Package.resolved
@@ -29,21 +29,21 @@
         }
       },
       {
-        "package": "CPostgreSQL",
-        "repositoryURL": "https://github.com/vapor-community/cpostgresql.git",
+        "package": "Core",
+        "repositoryURL": "https://github.com/vapor/core.git",
         "state": {
           "branch": null,
-          "revision": "33f0f94f023ab61ccd0ba2f076dd6a448be853e3",
-          "version": "2.1.0"
+          "revision": "b8330808f4f6b69941961afe8ad6b015562f7b7c",
+          "version": "2.1.2"
         }
       },
       {
-        "package": "Core",
-        "repositoryURL": "https://github.com/vapor/core.git",
+        "package": "CPostgreSQL",
+        "repositoryURL": "https://github.com/vapor-community/cpostgresql.git",
         "state": {
           "branch": null,
-          "revision": "b8330808f4f6b69941961afe8ad6b015562f7b7c",
-          "version": "2.1.2"
+          "revision": "33f0f94f023ab61ccd0ba2f076dd6a448be853e3",
+          "version": "2.1.0"
         }
       },
       {
@@ -65,21 +65,21 @@
         }
       },
       {
-        "package": "HTMLString",
-        "repositoryURL": "https://github.com/alexaubry/HTMLString",
+        "package": "HeliumLogger",
+        "repositoryURL": "https://github.com/IBM-Swift/HeliumLogger.git",
         "state": {
           "branch": null,
-          "revision": "f6cd285dce271ca2e22ffc5206174a2900a26d7c",
-          "version": "3.0.0"
+          "revision": "2afeb4c02f5a48a7fbd081fe8607b91eeebd7ecb",
+          "version": "1.7.1"
         }
       },
       {
-        "package": "HeliumLogger",
-        "repositoryURL": "https://github.com/IBM-Swift/HeliumLogger.git",
+        "package": "HTMLString",
+        "repositoryURL": "https://github.com/alexaubry/HTMLString",
         "state": {
           "branch": null,
-          "revision": "2afeb4c02f5a48a7fbd081fe8607b91eeebd7ecb",
-          "version": "1.7.1"
+          "revision": "f6cd285dce271ca2e22ffc5206174a2900a26d7c",
+          "version": "3.0.0"
         }
       },
       {
@@ -92,21 +92,21 @@
         }
       },
       {
-        "package": "Kitura-TemplateEngine",
-        "repositoryURL": "https://github.com/IBM-Swift/Kitura-TemplateEngine.git",
+        "package": "Kitura-net",
+        "repositoryURL": "https://github.com/IBM-Swift/Kitura-net.git",
         "state": {
           "branch": null,
-          "revision": "8e75a213bf4dd7b5effd03848335e344535969aa",
-          "version": "1.7.2"
+          "revision": "c29812e62855d34d32305e1dc06d2f1df5224df7",
+          "version": "1.7.19"
         }
       },
       {
-        "package": "Kitura-net",
-        "repositoryURL": "https://github.com/IBM-Swift/Kitura-net.git",
+        "package": "Kitura-TemplateEngine",
+        "repositoryURL": "https://github.com/IBM-Swift/Kitura-TemplateEngine.git",
         "state": {
           "branch": null,
-          "revision": "c29812e62855d34d32305e1dc06d2f1df5224df7",
-          "version": "1.7.19"
+          "revision": "8e75a213bf4dd7b5effd03848335e344535969aa",
+          "version": "1.7.2"
         }
       },
       {
@@ -173,39 +173,39 @@
         }
       },
       {
-        "package": "SQLite",
-        "repositoryURL": "https://github.com/vapor/sqlite.git",
+        "package": "Signals",
+        "repositoryURL": "https://github.com/IBM-Swift/BlueSignals.git",
         "state": {
           "branch": null,
-          "revision": "9aceb6a0d7b1698a557647493bd78b030dad468b",
-          "version": "2.3.1"
+          "revision": "0fab5a9f166e96f17bb9646dc56cd22a6f3c3757",
+          "version": "0.9.50"
         }
       },
       {
-        "package": "SSLService",
-        "repositoryURL": "https://github.com/IBM-Swift/BlueSSLService.git",
+        "package": "Socket",
+        "repositoryURL": "https://github.com/IBM-Swift/BlueSocket.git",
         "state": {
           "branch": null,
-          "revision": "7adf7cacc447d447df5dcc3058cabd402177c5aa",
-          "version": "0.12.64"
+          "revision": "00b126a2c19520bbf51cd26e8c0cc41dacbc5a05",
+          "version": "0.12.76"
         }
       },
       {
-        "package": "Signals",
-        "repositoryURL": "https://github.com/IBM-Swift/BlueSignals.git",
+        "package": "SQLite",
+        "repositoryURL": "https://github.com/vapor/sqlite.git",
         "state": {
           "branch": null,
-          "revision": "0fab5a9f166e96f17bb9646dc56cd22a6f3c3757",
-          "version": "0.9.50"
+          "revision": "9aceb6a0d7b1698a557647493bd78b030dad468b",
+          "version": "2.3.1"
         }
       },
       {
-        "package": "Socket",
-        "repositoryURL": "https://github.com/IBM-Swift/BlueSocket.git",
+        "package": "SSLService",
+        "repositoryURL": "https://github.com/IBM-Swift/BlueSSLService.git",
         "state": {
           "branch": null,
-          "revision": "00b126a2c19520bbf51cd26e8c0cc41dacbc5a05",
-          "version": "0.12.76"
+          "revision": "7adf7cacc447d447df5dcc3058cabd402177c5aa",
+          "version": "0.12.64"
         }
       },
       {
```

Reply:
Some of these even appear to have *downgraded* :'(

Reply:
Grabbing new versions of libraries is not silent. When dependencies are resolved (using `swift package resolve` which is called by `swift build`), versions are stored in `Package.resolved`. Subsequent `swift package resolve` runs first attempt to use the resolved versions, only using other versions if they are no longer valid. 
To update packages, we run the `swift package update` command.
It looks like your `Package.resolved` has not been changed, it's just been reordered. We have seen this before and I'm not entirely sure why it happens. Could you try running again and seeing if it changes again?

More detail [here](https://github.com/apple/swift-evolution/blob/master/proposals/0175-package-manager-revised-dependency-resolution.md) (unfortunately, the evolution proposals are the most detailed documentation available)

Reply:
Should I push my changes to `Package.resolved` then?

Reply:
These changes happened after running `docker-compose up`

Reply:
Your other issue might have been to do with you existing .build folder. Maybe try resetting, ditching the build folder, and running `docker-compose up` (which calls `swift build` fyi)

Reply:
.upToNextMinor(from: "x.y.z") means any x.*.* greater than x.y.z

Reply:
That worked!

Reply:
Did .resolved still change?

Reply:
No, it didn't.

Reply:
Interesting!

Reply:
So delete.  . build before using this!